packages/common/src/api/tan-query/wallets/useUserCoinBalance.ts

```diff
@@ -0,0 +1,88 @@
+import { useQuery } from '@tanstack/react-query'
+
+import { useCurrentUserId } from '~/api/tan-query/users/account/useCurrentUserId'
+import { ID } from '~/models/Identifiers'
+
+import { QUERY_KEYS } from '../queryKeys'
+import { QueryKey, SelectableQueryOptions } from '../types'
+
+type UserCoinBalanceResponse = {
+  data: Array<{
+    owner: string
+    account: string
+    balance: number
+  }>
+}
+
+export const getUserCoinBalanceQueryKey = (
+  userId: ID | null | undefined,
+  mint: string | null | undefined
+) =>
+  [
+    QUERY_KEYS.userCoinBalance,
+    userId,
+    mint
+  ] as unknown as QueryKey<UserCoinBalanceResponse | null>
+
+type UseUserCoinBalanceParams = {
+  userId: ID | null | undefined
+  mint: string
+}
+
+/**
+ * Hook to get a user's balance for a specific coin mint address.
+ * Currently mocked since the SDK method is not yet implemented.
+ * For the current user, always returns a balance.
+ * For other users, returns a balance 50% of the time.
+ */
+export const useUserCoinBalance = <TResult = UserCoinBalanceResponse | null>(
+  { userId, mint }: UseUserCoinBalanceParams,
+  options?: SelectableQueryOptions<UserCoinBalanceResponse | null, TResult>
+) => {
+  const { data: currentUserId } = useCurrentUserId()
+  const validUserId = !!userId && userId > 0
+  const validMint = !!mint
+
+  return useQuery({
+    queryKey: getUserCoinBalanceQueryKey(userId, mint),
+    queryFn: async (): Promise<UserCoinBalanceResponse | null> => {
+      const isCurrentUser = userId === currentUserId
+
+      // For current user, always return a balance
+      if (isCurrentUser) {
```

Comment:
do you think these user balances should be a part of user object at some point? this hook is still great and can be formatted to to check in user-cache first too, just thinking ahead

Reply:
i had same thought but that won't rly scale. just going off of the feature spec doc API.